File: inference/ch3-compute/benchmark_gemm.py
import time

import torch


def benchmark_gemm(m, n, k):
    a = torch.randn(m, k, dtype=torch.float32, device='cuda')
    b = torch.randn(k, n, dtype=torch.float32, device='cuda')
    torch.cuda.synchronize()
    start = time.perf_counter()
    for _ in range(100):
        _ = torch.matmul(a, b)
    torch.cuda.synchronize()
    return (2 * m * n * k * 100) / (time.perf_counter() - start) / 1e12


if __name__ == '__main__':
    for d in (512, 1024, 2048, 4096, 6144, 8192):
        perf = benchmark_gemm(d, d, d)
        print(f"{d}x{d}: {perf} TFLOPS")

File: inference/ch3-compute/benchmark_gemm.rs
use std::{hint::black_box, time::Instant};

use candle_core::Tensor;

fn benchmark_gemm(m: usize, n: usize, k: usize) -> candle_core::Result<f64> {
    let cuda = candle_core::Device::new_cuda(0)?;
    let a = Tensor::rand(0f32, 1.0, &[m, k], &cuda)?;
    let b = Tensor::rand(0f32, 1.0, &[k, n], &cuda)?;
    cuda.synchronize()?;
    let start = Instant::now();
    for _ in 0..100 {
        black_box(a.matmul(&b)?);
    }
    cuda.synchronize()?;
    Ok((2 * m * n * k * 100) as f64 / start.elapsed().as_secs_f64() / 1e12)
}

fn main() {
    for d in [512, 1024, 2048, 4096, 6144, 8192] {
        let perf = benchmark_gemm(d, d, d).unwrap();
        println!("{d}x{d}: {perf} TFLOPS");
    }
}
